NMOS registration implemented and functional

ctwarog-ross opened the pull request:
Closes issue #1025

Created NmosNode class that connects to an instance of Sony Nmos Registry
NmosNode discovers and registers both a node and device into the registry
Avahi libraries used are network setup agnostic, work with mdns and dns-sd
discovery_REST example device that creates instance of NmosNode

ctwarog-ross committed:
NMOS registration implemented and functional

Closes issue #1025

Created NmosNode class that connects to an instance of Sony Nmos Registry
NmosNode discovers and registers both a node and device into the registry
Avahi libraries used are network setup agnostic, work with mdns and dns-sd
discovery_REST example device that creates instance of NmosNode

ctwarog-ross committed:
fixed apt install prereqs for githib workflow

ctwarog-ross committed:
fixed apt install prereqs for githib workflow

ctwarog-ross committed:
fixed apt install prereqs for githib workflow

abrown-RV commented on .github/workflows/cpp-pull-request.yml:
gonna need to update the toolchain dockerfile too

abrown-RV replied:
https://github.com/rossvideo/Catena-deployments/issues/4

abrown-RV commented on sdks/cpp/common/include/NmosNode.h:
Have an interface class too, for mocking?

ctwarog-ross committed:
heartbeat code fix

abrown-RV commented on sdks/cpp/connections/REST/examples/discovery_REST/device.discovery.json:
don't forget the mandatory params... or maybe do forget them and let the codegen take care of it once Jason is done.

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+    // Compose minimal resources
+    std::string node_id = random_uuid();
+    std::string dev_id = random_uuid();
+    // TODO: detect local IP to reach your Node API; using 127.0.0.1 as placeholder
```
Friendly reminder machines can have multiple ips assigned. Might need some logic to decide which interface's IP to use.

abrown-RV replied:
I see you've got functions already, so nevermind!

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+
+    // Choose a registry
+    auto sel = choose_registry_and_build_base(8080);
+    if (!sel) { LOG(ERROR) << "No registry discovered. Exiting.\n"; return 2; }
```
Suggest well defined error codes with static consts defined on NmosNode

ctwarog-ross replied:
like throwing an enum code or something?

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+}
+
+std::string NmosNode::random_uuid() {
+    // NOT cryptographically strong. Good enough for example code.
```
Does it NEED to be cryptographically strong?

ctwarog-ross replied:
no, hence why I left it for now

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+}
+
+bool NmosNode::pick_primary_iface(std::string& ifname, std::string& mac){
+    DIR* d = opendir("/sys/class/net"); if (!d) return false;
```
consider c++17 std::filesystem? (I don't know if its any better AI Overview suggested it XD)

Would this not pick the first interface with `oper == "up"` in alphabetical order? I don't think that's necessarily the *primary*

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+        "\"tags\": {},"
+        "\"services\": ["
+            "{"
+                "\"href\": \"http://%s:%d/x-catena/status/\","
```
So hold on a sec.... is this a URL on the catena service that the registry expects to be able to call?

ctwarog-ross replied:
yes

abrown-RV replied:
So this node stuff needs to then make a whole http server. Listening on another port OR hook into our existing rest server for the REST half. I also suggest making a totally standalone container image for the node so we can sidecar it.

ctwarog-ross replied:
maybe we can have that as a separate issue, making PRs too big is less than ideal

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+    "{"
+        "\"id\": \"%s\","
+        "\"version\": \"%s\","
+        "\"label\": \"Catena Node\","
```
I feel like a bunch of these hardcoded values should be user configurable.

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+    return fmt(
+    "{"
+        "\"id\": \"%s\","
+        "\"version\": \"%s\","
```
weird that "version" is just the current time in a specific format, what are the nmos guys doing? XD

ctwarog-ross replied:
version is essentially the time signature of the last time the node made any config updates with the registry. It is expected in a seconds:nanoseconds format

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,426 @@
+    auto sel = choose_registry_and_build_base(8080);
+    if (!sel) { LOG(ERROR) << "No registry discovered. Exiting.\n"; return 2; }
+
+    const char* bearerChars = std::getenv("NMOS_BEARER");
```
Where does this come from?

ctwarog-ross replied:
NMOS_BEARER was implemented with the use case in mind that until we get an authorization service implemented,  you pass in an arbitrary bearer code as an env var during container creation

abrown-RV replied:
Oh I see later in the code, IF there's a bearer it'll use it.

abrown-RV commented on sdks/cpp/common/src/utils.cpp:
```diff
@@ -118,4 +118,10 @@ std::string catena::from_base64(const std::string& encoded) {
     return decoded;
+}
+
+std::string catena::fmt(const char* f, ...) {
```
Is it worth just grabbing a json library? Might make life easier. There are "header-only" libraries, that means there's no linking right? So no need for any extra `apt install`?

ctwarog-ross replied:
Im thinking at this point thats the best idea yeah

ctwarog-ross committed:
Added INmosNode interface
IP detection now chooses the ip on the same subnet as the registry address
node and device request now created with json library

ctwarog-ross committed:
fix

abrown-RV commented on sdks/cpp/common/include/INmosNode.h:
```diff
@@ -0,0 +1,124 @@
+
+    virtual std::string make_device_json(const std::string& dev_id, const std::string& node_id, int node_port) = 0;
+
+  protected:
```
I don't think protected should go on interfaces.

ctwarog-ross replied:
why not?

abrown-RV replied:
Nothing preventing it of course since they're not real interfaces, but interfaces describe the public ...interface that code uses to interact with the class. So hidden members may as well be left up to the implementors. For us I think we mainly do interfaces for mocking. And there's no point in mocking non public members.

abrown-RV commented on sdks/cpp/common/include/INmosNode.h:
```diff
@@ -0,0 +1,124 @@
+
+    virtual std::string make_node_json(const std::string& node_id, int node_port) = 0;
+
+    virtual std::string make_device_json(const std::string& dev_id, const std::string& node_id, int node_port) = 0;
```
These look like internal helper functions

ctwarog-ross replied:
you're right, I had them like this to give myself a blank slate for mocking. Bad design, removing from interface

abrown-RV commented on sdks/cpp/common/include/NmosNode.h:
```diff
@@ -0,0 +1,216 @@
+ * @file Authorizer.h
+ * @brief Class for handling authorization information
+ * @author John R. Naylor john.naylor@rossvideo.com
+ * @author John Danen
+ * @date 2024-09-18
+ * @author Ben Whitten benjamin.whitten@rossvideo.com
+ * @author Zuhayr Sarker zuhayr.sarker@rossvideo.com
+ * @date 2025-02-20
+ * @copyright Copyright (c) 2024 Ross Video
+ */
```
Clean up your copy paste XD

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+
+    DEBUG_LOG << "Avahi client created successfully.";
+
+    // Browse for _http._tcp services, change as needed
```
```suggestion
    // Browse for _nmos-registration._tcp services
```

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+    // Run discovery for a short window (e.g., 2 seconds) then proceed
+    auto start = std::chrono::steady_clock::now();
+    std::thread loop_thr([&](){ avahi_simple_poll_loop(simple_poll_); });
+    while (std::chrono::steady_clock::now() - start < std::chrono::seconds(2)) {
```
if you calculate now + 2 seconds before starting and compare against that. Instead of subtracting start from now. You'll save +-40 subtract operations. Don't know if we care XD

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+
+// ---------------------- Registration + heartbeat ----------------------
+
+std::optional<NmosNode::RegistrySelection> NmosNode::choose_registry_and_build_base(int node_port) {
```
`node_port` is not used.

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+
+    if (!get_node_iface(candidates_[0].addr)) return false;
+
+    json node_json = make_node_json(node_id, 8080);
```
This port 8080 just a hardcoded dummy for now? until we sort out the Registry -> Node stuff? Still might be good to variable it so its easy to update when we do figure it out.

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+    DEBUG_LOG << "Registered Node " << node_id << ". Heartbeating. Press Ctrl+C to exit.";
+
+    // Wait for signal
+    while (!stop_.load()) std::this_thread::sleep_for(std::chrono::milliseconds(200));
```
I think the idea is switch to the `catena::common::Heartbeat` which will work a little differently. But that's probably a good thing, cause won't this block the calling thread until `stop_ == true`? I feel like that's bad.

abrown-RV commented on sdks/cpp/common/include/INmosNode.h:
```diff
@@ -0,0 +1,124 @@
+
+    virtual std::vector<RegistryCandidate> getCandidates() = 0;
+
+    virtual AvahiClient* getClient() = 0;
```
Needs Doxygen comments

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,422 @@
+
+//figure out how to get subnet and match it to the current ip in the loop
+bool NmosNode::get_node_iface(const std::string& reg_addr){
+    std::string subnet = reg_addr.substr(0, reg_addr.rfind('.'));
```
subnet mask is not always /24. It's probably the most common, but should probably allow for other subnet masks.

ctwarog-ross replied:
have it in the arg with default value of some sort?

abrown-RV replied:
I assume its something you can determine from the network settings. Or just like take the IP with the highest number of matching bits or something...

abrown-RV commented on sdks/cpp/common/include/NmosNode.h:
```diff
@@ -0,0 +1,216 @@
+ * @copyright Copyright (c) 2024 Ross Video
+ */
+
+#include <curl/curl.h>
```
can this move to the cpp?

abrown-RV reviewed (commented):
Some comments, you were right! XD

ctwarog-ross committed:
Rough unit testing implementation

abrown-RV commented on unittests/cpp/common/tests/NmosNode_test.cpp:
```diff
@@ -0,0 +1,203 @@
+        inet_pton(AF_INET, "127.0.0.1", &a.sin_addr);
+        if (bind(listen_fd, (sockaddr*)&a, sizeof(a)) < 0) return false;
+        if (listen(listen_fd, 16) < 0) return false;
+        thr = std::thread([this]{
```
Suggest grabbing an http server lib if you ever want to mock more than a single endpoint (or even for the single to make it cleaner). Nathan found one for the web example

abrown-RV reviewed (commented):
Super cool trick with the --wrap. I guess you could also use to to totally swap out all of curl as well. But actually I think just spinning up a "mock" http server is actually less work, looking at all the different curl functions 😕 

This is still dropping the line coverage by at least a %, have to look at the nmosnode coverage to have ideas on covering more

ctwarog-ross committed:
Unit tests 94.4% coverage
Seperated init into seperate functions
init fails when functions called return false
Revamped unit tests for new design
Wrapped necessary avahi calls to force error bounds checking

ctwarog-ross committed:
unittests done

abrown-RV commented on .devcontainer/toolchain-cpp.dockerfile:
```diff
@@ -28,7 +28,10 @@ RUN . /root/toolchain.env \
-    curl=$CURL_VERSION \
+    curl=$CURL_VERSION libcurl4-openssl-dev\
+    libavahi-client-dev libavahi-common-dev \
+    nlohmann-json3-dev \
```
Does more version locking make sense for avahi or json?

ctwarog-ross replied:
I dont think at this time

abrown-RV commented on unittests/cpp/REST/CMakeLists.txt:
```diff
@@ -33,7 +33,6 @@ foreach(test_file ${REST_TEST_FILES})
     get_filename_component(test_name ${test_file} NAME_WE)
     add_executable(REST_${test_name} tests/${test_file})
     target_link_libraries(REST_${test_name} PRIVATE
-        catena_proto_common
```
Why?

ctwarog-ross replied:
not needed

abrown-RV commented on sdks/cpp/common/src/utils.cpp:
```diff
@@ -26,6 +26,12 @@
+    } catch (const std::exception& e) { 
+        LOG(ERROR) << "File does not exist: " << e.what(); 
+        return ""; 
+    }
```
Did you verify the other uses of this function can handle the functionality change?

ctwarog-ross replied:
There is one test for it, I changed from EXPECT_THROW to EXPECT_EQ. In the areas of the code that use it, they are just more robust since now you callers arent responsible for catching the error

abrown-RV replied:
`ServiceCredentials` popping up again. I mean, the `readFile` function is also used here. Does it still work with the changed `readFile` behaviour?

ctwarog-ross replied:
The only thing that comes to mind is if somewhere up the calling chain was handling an error from the file read if it didnt exist, that block of code will never be executed now.

abrown-RV replied:
Exactly, and also now code that maybe was once expecting to not run if the file didn't exist, might now run with the empty string. So that code needs to be able to not die if given `""`

abrown-RV commented on sdks/cpp/common/src/utils.cpp:
```diff
@@ -118,4 +124,10 @@ std::string catena::from_base64(const std::string& encoded) {
     return decoded;
+}
+
+std::string catena::fmt(const char* f, ...) {
```
Unit tests? kind of covered indirectly by all the other code, but still good to make some direct unit tests. @NR-RV agreed?

abrown-RV commented on sdks/cpp/common/src/utils.cpp:
```diff
@@ -26,6 +26,12 @@
+    try { 
+        if (!std::filesystem::exists(path)) return ""; 
+    } catch (const std::exception& e) { 
+        LOG(ERROR) << "File does not exist: " << e.what(); 
```
Not what the exception means.

abrown-RV replied:
[`std::filesystem::exists()`](https://en.cppreference.com/w/cpp/filesystem/exists.html) returns false if it doesn't exist. The exceptions are from other OS errors. So coverage doesn't hit the catch.

ctwarog-ross replied:
I wasn't thinking when writing this, you're right needs rewrite

abrown-RV commented on unittests/cpp/common/tests/NmosNode_test.cpp:
```diff
@@ -0,0 +1,319 @@
+class NmosNodeTest : public ::testing::Test {
+protected:
+    void SetUp() override {
+        g_avahi_test_control.reset();
```
glog init? and teardown?

abrown-RV commented on unittests/cpp/common/mocks/WrapAvahiClient.h:
Would it be possible to still use mocks here?

the `MOCK_METHOD` macro does seem to resolve to normal function definition, albeit with some weird templating around the return and param types. That could mess with the wrap linking part.

But if that doesn't work could you make it so the struct is a collection of the `MOCK_METHOD`s and then the `extern "C"` part just calls those mocks? Then you could do all the regular and excellent `EXPECT_CALL` awesomeness.

ctwarog-ross replied:
The avahi lib canot be mocked as the functions are not virtual. The AvahiTestControl struct could contain mocked versions of the avahi classes, unsure what I would get from that though as theyre mostly structs that just contain data. Is that what you meant?

abrown-RV commented on unittests/cpp/common/tests/NmosNode_test.cpp:
```diff
@@ -0,0 +1,319 @@
+
+    EXPECT_FALSE(node.init());
+}
+
```
Throw in one with `AVAHI_BROWSER_REMOVE` `AVAHI_BROWSER_CACHE_EXHAUSTED` or `AVAHI_BROWSER_ALL_FOR_NOW` for free extra coverage in `browse_cb` for those other switch cases

abrown-RV commented on unittests/cpp/common/tests/NmosNode_test.cpp:
```diff
@@ -0,0 +1,319 @@
+
+    reg.stop_and_join();
+}
+
```
- Add a manual call to `resolve_cb` with `event != AVAHI_RESOLVER_FOUND` to get free coverage on the else of the if in `resolve_cb`.
- Add a test with a couple of manual calls to `resolve_cb` with different candidates to test the sorting of candidates and get coverage on the sort callback. Could even be a couple of tests to cover all the permutations of the sorting logic.

abrown-RV commented on unittests/cpp/common/tests/NmosNode_test.cpp:
```diff
@@ -0,0 +1,319 @@
+                std::string method = buf.substr(0, sp1);
+                std::string path = buf.substr(sp1 + 1, sp2 - sp1 - 1);
+
+                if (method == "POST" && path == "/x-nmos/registration/v1.3/resource") {
```
If this could succeed on the first call, and then fail on the 2nd we could cover the 2nd post failure in `sendRequests`. There some way to get a mock in here? It supports that behaviour, AND counting number of invokes.

ctwarog-ross replied:
is that a case worth covering? Its alot of wizardry to confirm handling of very simple logging and return code

abrown-RV replied:
true.... but 100% coverage! XD

abrown-RV reviewed (changes requested):
Hopefully this doesn't hide my earlier comments. That wrap linker thing is super cool. Some comments on getting to 100% coverage in `NmosNode.cpp`, I think its doable.

ctwarog-ross committed:
You're up Andrew

ctwarog-ross committed:
You're up fr now Andrew

abrown-RV commented on sdks/cpp/common/src/NmosNode.cpp:
```diff
@@ -0,0 +1,451 @@
+    return std::nullopt;
+}
+
+void NmosNode::run_heartbeat(std::string base, std::string node_id, std::string bearer, std::chrono::milliseconds interval) {
```
Not using the Heartbeat class?

ctwarog-ross replied:
No9 we went over it, I had implemented it but it meant NmosNode was more complicated to set up and not self contained. Heartbeat is part of the Nmos spec so I thought it best to keep that self contained

abrown-RV reviewed (commented):
Not too sure where the segfault is coming from. Maybe using the heartbeat class will fix it?

ctwarog-ross committed:
Uses implemented Heartbeast class
Fixed shutdown sequence
Removed unnecessary package installs
Version locked libraries in the toolchain

ctwarog-ross committed:
fixed doxygen and copyright blocks

abrown-RV reviewed (approved):
💯
